compiler/rustc_ast/src/ast.rs

```diff
@@ -3185,9 +3185,13 @@ mod size_asserts {
     static_assert_size!(Impl, 136);
     static_assert_size!(Item, 136);
     static_assert_size!(ItemKind, 64);
-    static_assert_size!(LitKind, 24);
+    // This can be removed after i128:128 is in the bootstrap compiler's target.
+    #[cfg(not(bootstrap))]
+    static_assert_size!(LitKind, 32);
     static_assert_size!(Local, 72);
-    static_assert_size!(MetaItemLit, 40);
+    // This can be removed after i128:128 is in the bootstrap compiler's target.
+    #[cfg(not(bootstrap))]
+    static_assert_size!(MetaItemLit, 48);
```

Comment:
I don't think config on bootstrap is too reliable since we could wind up with a bootstrap with either LLVM version. It would be better if there is a way to check the LLVM version here.

Alternatively, a static assertion that size can be either 24 or 32 (or 40/48) with a `// FIXME:` is probably suitable

Reply:
This doesn't depend on the LLVM version - it depends on the alignment in the spec file. If you refer back to the [zulip conversation](https://rust-lang.zulipchat.com/#narrow/stream/187780-t-compiler.2Fwg-llvm/topic/LLVM.20alignment.20of.20i128) we had, you'll note that `TargetDataLayout` will determine the size of these types, which is derived from the pre-munged `data_layout`, not the one after correction.

So in theory, after this rev, this should always be the case, regardless of LLVM version linked.

All that said, I might be misunderstanding something here, because I'm getting some creader issues in stage2+

Reply:
Hm, then I guess those are coming from rustc using the new layout strings (16-byte align) while llvm <18 uses the adjusted version (8byte)? What happens if you update the layout strings, don't do the LLVM<18 fixup, and just suppress the "different default layout strings" error?

Reply:
I've just tested, and that works (`./x.py test` and `./x.py dist` both complete locally).

The check still seemed valuable, so I've uploaded a change that makes the `bug!` conditional check the munged data layout rather than munging it when sending it to LLVM.

Reply:
You can't do this, it will cause LLVM to assert. Even if LLVM didn't assert it would break cross-language LTO.

Reply:
You seem to be right. I must have had asserts off locally or something, because this *did* work with the vendored LLVM locally.

Given that:
1. We can't adjust rustc's data layout without adjusting LLVM's data layout, that results in bugs.
2. We can't adjust LLVM's data layout on older LLVM.

Does that just leave the "newer LLVM needs to munge the `data_layout` early in execution, before `TargetDataLayout` gets built?